src/arch/LowLevelWindow/LowLevelWindow_SDL.cpp

```diff
@@ -0,0 +1,258 @@
+#include "global.h"
+#include "LowLevelWindow_SDL.h"
+#include "arch/arch_default.h"
+#include "RageLog.h"
+
+#include <SDL3/SDL.h>
+
+#include "DisplaySpec.h"
+#include "Preference.h"
+#include "RageSurface.h"
+#include "RageSurface_Load.h"
+
+Preference<bool> g_bSDLLinuxUseWayland("SDLLinuxUseWayland", true);
+
+LowLevelWindow_SDL::LowLevelWindow_SDL()
+{
+    LOG->Info("[LLW_SDL] Using LowLevelWindow_SDL");
+    m_pWindow = nullptr;
+    m_pGlContext = nullptr;
+
+#if defined(LINUX)
+    // Nudge SDL3 to use wayland if requested (this is only needed until SDL uses wayland by default)
+    if (g_bSDLLinuxUseWayland.Get())
+    {
+        SDL_SetHint(SDL_HINT_VIDEO_DRIVER, "wayland");
+    }
+#endif
+
+    SDL_Init(SDL_INIT_VIDEO | SDL_INIT_EVENTS | SDL_INIT_JOYSTICK | SDL_INIT_GAMEPAD);
+}
+
+LowLevelWindow_SDL::~LowLevelWindow_SDL()
+{
+    SDL_GL_DestroyContext(m_pGlContext);
+    SDL_DestroyWindow(m_pWindow);
+    SDL_Quit();
+}
+
+void* LowLevelWindow_SDL::GetProcAddress(RString s)
+{
+    return SDL_GL_GetProcAddress(s);
+}
+
+RString LowLevelWindow_SDL::TryVideoMode(const VideoModeParams& p, bool& bNewDeviceOut)
+{
+    // Ensure we're using compat profile
+    SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_COMPATIBILITY);
+
+    if (!m_pWindow)
+    {
+        m_pWindow = SDL_CreateWindow(p.sWindowTitle, p.width, p.height, SDL_WINDOW_OPENGL | SDL_WINDOW_HIDDEN | SDL_WINDOW_HIGH_PIXEL_DENSITY);
+        if (!m_pWindow)
+        {
+            LOG->Info("[LLW_SDL::TryVideoMode] Failed to create SDL Window: %s", SDL_GetError());
+            return "Failed to create SDL Window - Check logs";
+        }
+
+        SDL_SetWindowPosition(m_pWindow, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED);
+        m_pGlContext = SDL_GL_CreateContext(m_pWindow);
+
+        if (!m_pGlContext)
+        {
+            LOG->Info("[LLW_SDL::TryVideoMode] Failed to create GL Context: %s", SDL_GetError());
+            return "Failed to create GL Context - Check logs";
+        }
+
+        SDL_GL_MakeCurrent(m_pWindow, m_pGlContext);
+        SDL_ShowWindow(m_pWindow);
+
+        // New window created
+        bNewDeviceOut = true;
+    } else
+    {
+        bNewDeviceOut = false;
+    }
+
+    // Adjust for any hidpi scaling & prevent division by zero
+    // Note: This is rounded to the nearest 4 decimals to prevent improperly set scale values
+    auto windowScale = std::floor(SDL_GetWindowDisplayScale(m_pWindow) * 1000) / 1000;
```

Comment:
My dpi scale had some junk values past the 6th decimal place. I think it's safe to round here.